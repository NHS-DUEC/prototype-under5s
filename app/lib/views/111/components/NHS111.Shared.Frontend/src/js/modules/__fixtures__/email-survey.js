export const formHtml = `<form class="nhsuk-u-margin-top-3" method="post" action="/Survey/EmailSurveyCollection" data-survey-form="" target="_blank">
<input name="data-1" value="value-1">
<input name="data-2" value="value-2">

<div class="nhsuk-form-group nhsuk-u-margin-bottom-0" data-survey-form-section="">
    <p>We would like to know more about your experience using 111 online.</p>

    <p>We will send you a link to a survey in 2 days.</p>

    <p>This is so you can get the help you need before completing the survey.</p>

    <fieldset class="nhsuk-fieldset">
        <div class="nhsuk-form-group">
            <label class="nhsuk-label nhsuk-u-font-weight-bold" for="email">
                Enter your email address
            </label>
            <input class="nhsuk-input nhsuk-u-width-one-half" id="email" name="EmailAddress" type="email" inputmode="email">
        </div>
    </fieldset>

    <button class="nhsuk-button nhsuk-u-margin-top-4" data-email-survey-cancel type="submit">Send</button>
    <button class="nhsuk-button nhsuk-button--secondary nhsuk-u-margin-top-4" data-email-survey-submit type="submit">Send</button>
    <p data-feedback-nojs-message>(Opens in a new tab or window)</p>
</div>
</form>`;
