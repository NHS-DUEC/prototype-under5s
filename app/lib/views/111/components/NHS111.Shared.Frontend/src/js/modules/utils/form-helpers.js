export function hideNoJsMessage(form) {
  Array.from(form.querySelectorAll("[data-feedback-nojs-message]")).forEach(
    (item) => item.remove()
  );
}

export function clearErrors(form) {
  const errors = form.querySelectorAll("[data-feedback-error]");
  const formGroups = form.querySelectorAll(".nhsuk-form-group");
  const elementsWithOriginalAria = form.querySelectorAll(
    "[data-original-aria-describedby]"
  );

  Array.from(errors).forEach((item) => item && item.remove());
  Array.from(formGroups).forEach(
    (item) => item && item.classList.remove("nhsuk-form-group--error")
  );
  Array.from(elementsWithOriginalAria).forEach((el) =>
    el.setAttribute("aria-describedBy", el.dataset.originalAriaDescribedby)
  );
}

export function hideAllFormSections(form, formSections) {
  clearErrors(form);

  Array.from(formSections).forEach((element) => {
    element.hidden = true;
  });
}

export function trackOriginalAria(form) {
  // Keep track of any original aria-describedBy attributes so we can restore them when clearing errors
  Array.from(form.querySelectorAll("fieldset, input, textarea")).forEach(
    (el) => {
      if (el.getAttribute("aria-describedBy") !== null) {
        el.setAttribute(
          "data-original-aria-describedby",
          el.getAttribute("aria-describedBy")
        );
      }
    }
  );
}

export function formError(
  form,
  ErrorSummary,
  fieldId,
  errorMessage,
  errorLink = ""
) {
  const template = document.createElement("template");

  clearErrors(form);

  template.innerHTML = `
    <div class="nhsuk-error-summary" data-feedback-error data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#${fieldId}">${errorLink || errorMessage}</a>
          </li>
        </ul>
      </div>
    </div>`;

  const previousErrorSummary = form.querySelector(".nhsuk-error-summary");
  if (previousErrorSummary) {
    previousErrorSummary.remove();
  }

  const errorSummary = template.content;

  form.prepend(errorSummary);
  ErrorSummary();

  const field = form.querySelector(`#${fieldId}`);
  const formGroup = field.closest(".nhsuk-form-group");

  const errorMessageTemplate = document.createElement("template");
  errorMessageTemplate.innerHTML = `<span data-feedback-error data-test-id="feedback-error-message" class="nhsuk-error-message" id="${fieldId}-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> ${errorMessage}
    </span>`;

  if (field.getAttribute("type") === "radio") {
    formGroup.querySelector("legend").after(errorMessageTemplate.content);
    const fieldset = formGroup.querySelector("fieldset");
    fieldset.setAttribute(
      "aria-describedby",
      `${
        fieldset.dataset.originalAriaDescribedby
          ? fieldset.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  } else {
    formGroup.querySelector("label").after(errorMessageTemplate.content);
    field.setAttribute(
      "aria-describedby",
      `${
        field.dataset.originalAriaDescribedby
          ? field.dataset.originalAriaDescribedby
          : ""
      } ${fieldId}-error`
    );
  }

  formGroup.classList.add("nhsuk-form-group--error");
}
