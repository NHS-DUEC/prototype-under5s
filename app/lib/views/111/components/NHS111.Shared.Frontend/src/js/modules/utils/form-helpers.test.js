import ErrorSummary from "nhsuk-frontend/packages/components/error-summary/error-summary";

import { formHtml } from "../__fixtures__/feedback";

import * as formHelpers from "./form-helpers";

describe("hideNoJsMessage method", () => {
  test("hides any no-js messages in the form", () => {
    document.body.innerHTML = `
        <form>
          <p data-feedback-nojs-message>no-JS paragraph</p>
          <p>Normal paragraph</p>
          <p data-feedback-nojs-message>no-JS paragraph</p>
        </form>
      `;

    const feedbackForm = document.querySelector("form");

    formHelpers.hideNoJsMessage(feedbackForm);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe("clearErrors method", () => {
  test("restores html to original state", () => {
    document.body.innerHTML = `
      <form>
        <div data-feedback-error></div>
        <div class="nhsuk-form-group nhsuk-form-group--error"></div>
        <div data-feedback-error></div>
        <div class="nhsuk-form-group nhsuk-form-group--error"></div>
        <div data-original-aria-describedby="foo"></div>
        <div data-original-aria-describedby="bar"></div>
      </form>
      `;

    const feedbackForm = document.querySelector("form");

    formHelpers.clearErrors(feedbackForm);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe("hideAllFormSections method", () => {
  test("hides all form sections", () => {
    document.body.innerHTML = `
      <form>
        <div data-feedback-form-section></div>
        <div data-feedback-form-section></div>
      </form>
      `;

    const feedbackForm = document.querySelector("form");
    const feedbackFormSections = feedbackForm.querySelectorAll(
      "[data-feedback-form-section]"
    );

    formHelpers.hideAllFormSections(feedbackForm, feedbackFormSections);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe("formError method", () => {
  test("correctly sets errors on radio fields", () => {
    document.body.innerHTML = formHtml;

    const feedbackForm = document.querySelector("form");

    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-choice-yes",
      "test error message"
    );

    expect(document.body.innerHTML).toMatchSnapshot();

    expect(document.activeElement.dataset.testId).toEqual(
      "feedback-error-summary"
    );
  });

  test("correctly sets repeated errors on radio fields", () => {
    document.body.innerHTML = formHtml;

    const feedbackForm = document.querySelector("form");

    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-choice-yes",
      "test error message"
    );
    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-choice-yes",
      "a second error message"
    );
    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-choice-yes",
      "a third error message"
    );

    expect(document.body.innerHTML).toMatchSnapshot();

    expect(document.activeElement.dataset.testId).toEqual(
      "feedback-error-summary"
    );
  });

  test("correctly sets errors on textarea field", () => {
    document.body.innerHTML = formHtml;

    const feedbackForm = document.querySelector("form");

    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-text",
      "test error message"
    );

    expect(document.body.innerHTML).toMatchSnapshot();

    expect(document.activeElement.dataset.testId).toEqual(
      "feedback-error-summary"
    );
  });

  test("correctly sets repeated errors on textarea fields", () => {
    document.body.innerHTML = formHtml;

    const feedbackForm = document.querySelector("form");

    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-text",
      "test error message"
    );
    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-text",
      "a second error message"
    );
    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-text",
      "a third error message"
    );

    expect(document.body.innerHTML).toMatchSnapshot();

    expect(document.activeElement.dataset.testId).toEqual(
      "feedback-error-summary"
    );
  });

  test("correctly sets errors on textarea field with different error link relative to message", () => {
    document.body.innerHTML = formHtml;

    const feedbackForm = document.querySelector("form");

    formHelpers.formError(
      feedbackForm,
      ErrorSummary,
      "feedback-text",
      "test error message",
      "test error link"
    );

    expect(document.body.innerHTML).toMatchSnapshot();

    expect(document.activeElement.dataset.testId).toEqual(
      "feedback-error-summary"
    );
  });
});

describe("trackOriginalAria method", () => {
  test("stores original aria-describedby attributes on a temporary data attribute for later use", () => {
    document.body.innerHTML = `
        <form>
          <fieldset aria-describedby="fieldset">
            <input aria-describedby="input">
            <textarea aria-describedby="textarea"></textarea>
          </fieldset>

          <fieldset aria-describedby="fieldset2">
            <input aria-describedby="input2">
            <textarea aria-describedby="textarea2"></textarea>
          </fieldset>
        </form>
      `;

    const feedbackForm = document.querySelector("form");

    formHelpers.trackOriginalAria(feedbackForm);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
