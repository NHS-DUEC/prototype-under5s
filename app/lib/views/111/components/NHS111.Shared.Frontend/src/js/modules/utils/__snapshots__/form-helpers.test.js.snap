// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clearErrors method restores html to original state 1`] = `
"
      <form>
        
        <div class="nhsuk-form-group"></div>
        
        <div class="nhsuk-form-group"></div>
        <div data-original-aria-describedby="foo" aria-describedby="foo"></div>
        <div data-original-aria-describedby="bar" aria-describedby="bar"></div>
      </form>
      "
`;

exports[`formError method correctly sets errors on radio fields 1`] = `
"<form data-test-id="feedback-content" action="/Feedback/SubmitFeedback" method="post" target="_blank" data-feedback-form="" autocomplete="off">
    <div class="nhsuk-error-summary" data-feedback-error="" data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#feedback-choice-yes">test error message</a>
          </li>
        </ul>
      </div>
    </div>
<input type="hidden" name="UserId" value="c0604133-7150-442d-afbe-957f19902546">

<input type="hidden" name="PageData.Page" value="ModuleZero">
<input type="hidden" name="PageData.TxNumber" value="">
<input type="hidden" name="PageData.QuestionId" value="">
<input type="hidden" name="PageData.StartingPathwayNo" value="">
<input type="hidden" name="PageData.StartingPathwayTitle" value="">
<input type="hidden" name="PageData.Gender" value="Indeterminate">
<input type="hidden" name="PageData.Age" value="">
<input type="hidden" name="PageData.PathwayNo" value="">
<input type="hidden" name="PageData.PathwayTitle" value="">
<input type="hidden" name="PageData.DxCode" value="">
<input type="hidden" name="PageData.SearchString" value="">
<input type="hidden" name="PageData.Campaign" value="">
<input type="hidden" name="PageData.Source" value="">
<input type="hidden" name="PageData.StartUrl" value="default">

<div class="nhsuk-form-group nhsuk-form-group--error" data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset" aria-describedby=" feedback-choice-yes-error">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">
      Does the page make sense?
    </legend><span data-feedback-error="" data-test-id="feedback-error-message" class="nhsuk-error-message" id="feedback-choice-yes-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> test error message
    </span>
    <div class="nhsuk-radios">
      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-yes" name="feedbackChoice" type="radio" value="yes">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-yes">Yes</label>
      </div>

      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-no" name="feedbackChoice" type="radio" value="no">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-no">No</label>
      </div>
    </div>
  </fieldset>
</div>

<div data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m nhsuk-u-margin-bottom-0" data-test-id="feedback-explanation-1">
      Tell us how to make this page easier to understand
    </legend>

    <br>

    <div class="nhsuk-character-count" data-module="nhsuk-character-count" data-maxlength="1200">
      <div class="nhsuk-form-group">
        <div class="app-callout app-callout--info" id="feedback-contact-warning">
          <p data-test-id="feedback-explanation-2">
            Do <strong>not</strong> leave any personal or contact information.
            No one will reply to requests for medical help or contact you.
          </p>
        </div>

        <label for="feedback-text" class="nhsuk-label">Your feedback</label>
        <textarea class="nhsuk-textarea nhsuk-js-character-count" id="feedback-text" name="Text" rows="5" data-test-id="feedback-textarea" aria-describedby="feedback-text-info" data-autosize=""></textarea>
      </div>

      <div class="nhsuk-hint nhsuk-character-count__message" id="feedback-text-info">
        You can enter up to 1200 characters
      </div>
    </div>
  </fieldset>

  <button data-test-id="send-feedback-button" data-event-trigger="click" data-event-value="Send feedback" class="nhsuk-button nhsuk-u-margin-bottom-2" type="submit">
    Send feedback
  </button>
  <p class="nhsuk-hint" data-feedback-nojs-message="">
    (Opens in a new tab or window)
  </p>
</div>
</form>"
`;

exports[`formError method correctly sets errors on textarea field 1`] = `
"<form data-test-id="feedback-content" action="/Feedback/SubmitFeedback" method="post" target="_blank" data-feedback-form="" autocomplete="off">
    <div class="nhsuk-error-summary" data-feedback-error="" data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#feedback-text">test error message</a>
          </li>
        </ul>
      </div>
    </div>
<input type="hidden" name="UserId" value="c0604133-7150-442d-afbe-957f19902546">

<input type="hidden" name="PageData.Page" value="ModuleZero">
<input type="hidden" name="PageData.TxNumber" value="">
<input type="hidden" name="PageData.QuestionId" value="">
<input type="hidden" name="PageData.StartingPathwayNo" value="">
<input type="hidden" name="PageData.StartingPathwayTitle" value="">
<input type="hidden" name="PageData.Gender" value="Indeterminate">
<input type="hidden" name="PageData.Age" value="">
<input type="hidden" name="PageData.PathwayNo" value="">
<input type="hidden" name="PageData.PathwayTitle" value="">
<input type="hidden" name="PageData.DxCode" value="">
<input type="hidden" name="PageData.SearchString" value="">
<input type="hidden" name="PageData.Campaign" value="">
<input type="hidden" name="PageData.Source" value="">
<input type="hidden" name="PageData.StartUrl" value="default">

<div class="nhsuk-form-group" data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">
      Does the page make sense?
    </legend>
    <div class="nhsuk-radios">
      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-yes" name="feedbackChoice" type="radio" value="yes">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-yes">Yes</label>
      </div>

      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-no" name="feedbackChoice" type="radio" value="no">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-no">No</label>
      </div>
    </div>
  </fieldset>
</div>

<div data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m nhsuk-u-margin-bottom-0" data-test-id="feedback-explanation-1">
      Tell us how to make this page easier to understand
    </legend>

    <br>

    <div class="nhsuk-character-count" data-module="nhsuk-character-count" data-maxlength="1200">
      <div class="nhsuk-form-group nhsuk-form-group--error">
        <div class="app-callout app-callout--info" id="feedback-contact-warning">
          <p data-test-id="feedback-explanation-2">
            Do <strong>not</strong> leave any personal or contact information.
            No one will reply to requests for medical help or contact you.
          </p>
        </div>

        <label for="feedback-text" class="nhsuk-label">Your feedback</label><span data-feedback-error="" data-test-id="feedback-error-message" class="nhsuk-error-message" id="feedback-text-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> test error message
    </span>
        <textarea class="nhsuk-textarea nhsuk-js-character-count" id="feedback-text" name="Text" rows="5" data-test-id="feedback-textarea" aria-describedby=" feedback-text-error" data-autosize=""></textarea>
      </div>

      <div class="nhsuk-hint nhsuk-character-count__message" id="feedback-text-info">
        You can enter up to 1200 characters
      </div>
    </div>
  </fieldset>

  <button data-test-id="send-feedback-button" data-event-trigger="click" data-event-value="Send feedback" class="nhsuk-button nhsuk-u-margin-bottom-2" type="submit">
    Send feedback
  </button>
  <p class="nhsuk-hint" data-feedback-nojs-message="">
    (Opens in a new tab or window)
  </p>
</div>
</form>"
`;

exports[`formError method correctly sets errors on textarea field with different error link relative to message 1`] = `
"<form data-test-id="feedback-content" action="/Feedback/SubmitFeedback" method="post" target="_blank" data-feedback-form="" autocomplete="off">
    <div class="nhsuk-error-summary" data-feedback-error="" data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#feedback-text">test error link</a>
          </li>
        </ul>
      </div>
    </div>
<input type="hidden" name="UserId" value="c0604133-7150-442d-afbe-957f19902546">

<input type="hidden" name="PageData.Page" value="ModuleZero">
<input type="hidden" name="PageData.TxNumber" value="">
<input type="hidden" name="PageData.QuestionId" value="">
<input type="hidden" name="PageData.StartingPathwayNo" value="">
<input type="hidden" name="PageData.StartingPathwayTitle" value="">
<input type="hidden" name="PageData.Gender" value="Indeterminate">
<input type="hidden" name="PageData.Age" value="">
<input type="hidden" name="PageData.PathwayNo" value="">
<input type="hidden" name="PageData.PathwayTitle" value="">
<input type="hidden" name="PageData.DxCode" value="">
<input type="hidden" name="PageData.SearchString" value="">
<input type="hidden" name="PageData.Campaign" value="">
<input type="hidden" name="PageData.Source" value="">
<input type="hidden" name="PageData.StartUrl" value="default">

<div class="nhsuk-form-group" data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">
      Does the page make sense?
    </legend>
    <div class="nhsuk-radios">
      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-yes" name="feedbackChoice" type="radio" value="yes">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-yes">Yes</label>
      </div>

      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-no" name="feedbackChoice" type="radio" value="no">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-no">No</label>
      </div>
    </div>
  </fieldset>
</div>

<div data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m nhsuk-u-margin-bottom-0" data-test-id="feedback-explanation-1">
      Tell us how to make this page easier to understand
    </legend>

    <br>

    <div class="nhsuk-character-count" data-module="nhsuk-character-count" data-maxlength="1200">
      <div class="nhsuk-form-group nhsuk-form-group--error">
        <div class="app-callout app-callout--info" id="feedback-contact-warning">
          <p data-test-id="feedback-explanation-2">
            Do <strong>not</strong> leave any personal or contact information.
            No one will reply to requests for medical help or contact you.
          </p>
        </div>

        <label for="feedback-text" class="nhsuk-label">Your feedback</label><span data-feedback-error="" data-test-id="feedback-error-message" class="nhsuk-error-message" id="feedback-text-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> test error message
    </span>
        <textarea class="nhsuk-textarea nhsuk-js-character-count" id="feedback-text" name="Text" rows="5" data-test-id="feedback-textarea" aria-describedby=" feedback-text-error" data-autosize=""></textarea>
      </div>

      <div class="nhsuk-hint nhsuk-character-count__message" id="feedback-text-info">
        You can enter up to 1200 characters
      </div>
    </div>
  </fieldset>

  <button data-test-id="send-feedback-button" data-event-trigger="click" data-event-value="Send feedback" class="nhsuk-button nhsuk-u-margin-bottom-2" type="submit">
    Send feedback
  </button>
  <p class="nhsuk-hint" data-feedback-nojs-message="">
    (Opens in a new tab or window)
  </p>
</div>
</form>"
`;

exports[`formError method correctly sets repeated errors on radio fields 1`] = `
"<form data-test-id="feedback-content" action="/Feedback/SubmitFeedback" method="post" target="_blank" data-feedback-form="" autocomplete="off">
    <div class="nhsuk-error-summary" data-feedback-error="" data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#feedback-choice-yes">a third error message</a>
          </li>
        </ul>
      </div>
    </div>
    
    
<input type="hidden" name="UserId" value="c0604133-7150-442d-afbe-957f19902546">

<input type="hidden" name="PageData.Page" value="ModuleZero">
<input type="hidden" name="PageData.TxNumber" value="">
<input type="hidden" name="PageData.QuestionId" value="">
<input type="hidden" name="PageData.StartingPathwayNo" value="">
<input type="hidden" name="PageData.StartingPathwayTitle" value="">
<input type="hidden" name="PageData.Gender" value="Indeterminate">
<input type="hidden" name="PageData.Age" value="">
<input type="hidden" name="PageData.PathwayNo" value="">
<input type="hidden" name="PageData.PathwayTitle" value="">
<input type="hidden" name="PageData.DxCode" value="">
<input type="hidden" name="PageData.SearchString" value="">
<input type="hidden" name="PageData.Campaign" value="">
<input type="hidden" name="PageData.Source" value="">
<input type="hidden" name="PageData.StartUrl" value="default">

<div class="nhsuk-form-group nhsuk-form-group--error" data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset" aria-describedby=" feedback-choice-yes-error">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">
      Does the page make sense?
    </legend><span data-feedback-error="" data-test-id="feedback-error-message" class="nhsuk-error-message" id="feedback-choice-yes-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> a third error message
    </span>
    <div class="nhsuk-radios">
      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-yes" name="feedbackChoice" type="radio" value="yes">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-yes">Yes</label>
      </div>

      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-no" name="feedbackChoice" type="radio" value="no">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-no">No</label>
      </div>
    </div>
  </fieldset>
</div>

<div data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m nhsuk-u-margin-bottom-0" data-test-id="feedback-explanation-1">
      Tell us how to make this page easier to understand
    </legend>

    <br>

    <div class="nhsuk-character-count" data-module="nhsuk-character-count" data-maxlength="1200">
      <div class="nhsuk-form-group">
        <div class="app-callout app-callout--info" id="feedback-contact-warning">
          <p data-test-id="feedback-explanation-2">
            Do <strong>not</strong> leave any personal or contact information.
            No one will reply to requests for medical help or contact you.
          </p>
        </div>

        <label for="feedback-text" class="nhsuk-label">Your feedback</label>
        <textarea class="nhsuk-textarea nhsuk-js-character-count" id="feedback-text" name="Text" rows="5" data-test-id="feedback-textarea" aria-describedby="feedback-text-info" data-autosize=""></textarea>
      </div>

      <div class="nhsuk-hint nhsuk-character-count__message" id="feedback-text-info">
        You can enter up to 1200 characters
      </div>
    </div>
  </fieldset>

  <button data-test-id="send-feedback-button" data-event-trigger="click" data-event-value="Send feedback" class="nhsuk-button nhsuk-u-margin-bottom-2" type="submit">
    Send feedback
  </button>
  <p class="nhsuk-hint" data-feedback-nojs-message="">
    (Opens in a new tab or window)
  </p>
</div>
</form>"
`;

exports[`formError method correctly sets repeated errors on textarea fields 1`] = `
"<form data-test-id="feedback-content" action="/Feedback/SubmitFeedback" method="post" target="_blank" data-feedback-form="" autocomplete="off">
    <div class="nhsuk-error-summary" data-feedback-error="" data-test-id="feedback-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1">
      <h2 class="nhsuk-error-summary__title" id="error-summary-title">
        There is a problem
      </h2>
      <div class="nhsuk-error-summary__body">
        <ul class="nhsuk-list nhsuk-error-summary__list">
          <li>
            <a href="#feedback-text">a third error message</a>
          </li>
        </ul>
      </div>
    </div>
    
    
<input type="hidden" name="UserId" value="c0604133-7150-442d-afbe-957f19902546">

<input type="hidden" name="PageData.Page" value="ModuleZero">
<input type="hidden" name="PageData.TxNumber" value="">
<input type="hidden" name="PageData.QuestionId" value="">
<input type="hidden" name="PageData.StartingPathwayNo" value="">
<input type="hidden" name="PageData.StartingPathwayTitle" value="">
<input type="hidden" name="PageData.Gender" value="Indeterminate">
<input type="hidden" name="PageData.Age" value="">
<input type="hidden" name="PageData.PathwayNo" value="">
<input type="hidden" name="PageData.PathwayTitle" value="">
<input type="hidden" name="PageData.DxCode" value="">
<input type="hidden" name="PageData.SearchString" value="">
<input type="hidden" name="PageData.Campaign" value="">
<input type="hidden" name="PageData.Source" value="">
<input type="hidden" name="PageData.StartUrl" value="default">

<div class="nhsuk-form-group" data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m">
      Does the page make sense?
    </legend>
    <div class="nhsuk-radios">
      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-yes" name="feedbackChoice" type="radio" value="yes">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-yes">Yes</label>
      </div>

      <div class="nhsuk-radios__item">
        <input class="nhsuk-radios__input" id="feedback-choice-no" name="feedbackChoice" type="radio" value="no">
        <label class="nhsuk-label nhsuk-radios__label" for="feedback-choice-no">No</label>
      </div>
    </div>
  </fieldset>
</div>

<div data-feedback-form-section="">
  <fieldset class="nhsuk-fieldset">
    <legend class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m nhsuk-u-margin-bottom-0" data-test-id="feedback-explanation-1">
      Tell us how to make this page easier to understand
    </legend>

    <br>

    <div class="nhsuk-character-count" data-module="nhsuk-character-count" data-maxlength="1200">
      <div class="nhsuk-form-group nhsuk-form-group--error">
        <div class="app-callout app-callout--info" id="feedback-contact-warning">
          <p data-test-id="feedback-explanation-2">
            Do <strong>not</strong> leave any personal or contact information.
            No one will reply to requests for medical help or contact you.
          </p>
        </div>

        <label for="feedback-text" class="nhsuk-label">Your feedback</label><span data-feedback-error="" data-test-id="feedback-error-message" class="nhsuk-error-message" id="feedback-text-error">
      <span class="nhsuk-u-visually-hidden">Error:</span> a third error message
    </span>
        <textarea class="nhsuk-textarea nhsuk-js-character-count" id="feedback-text" name="Text" rows="5" data-test-id="feedback-textarea" aria-describedby=" feedback-text-error" data-autosize=""></textarea>
      </div>

      <div class="nhsuk-hint nhsuk-character-count__message" id="feedback-text-info">
        You can enter up to 1200 characters
      </div>
    </div>
  </fieldset>

  <button data-test-id="send-feedback-button" data-event-trigger="click" data-event-value="Send feedback" class="nhsuk-button nhsuk-u-margin-bottom-2" type="submit">
    Send feedback
  </button>
  <p class="nhsuk-hint" data-feedback-nojs-message="">
    (Opens in a new tab or window)
  </p>
</div>
</form>"
`;

exports[`hideAllFormSections method hides all form sections 1`] = `
"
      <form>
        <div data-feedback-form-section="" hidden=""></div>
        <div data-feedback-form-section="" hidden=""></div>
      </form>
      "
`;

exports[`hideNoJsMessage method hides any no-js messages in the form 1`] = `
"
        <form>
          
          <p>Normal paragraph</p>
          
        </form>
      "
`;

exports[`trackOriginalAria method stores original aria-describedby attributes on a temporary data attribute for later use 1`] = `
"
        <form>
          <fieldset aria-describedby="fieldset" data-original-aria-describedby="fieldset">
            <input aria-describedby="input" data-original-aria-describedby="input">
            <textarea aria-describedby="textarea" data-original-aria-describedby="textarea"></textarea>
          </fieldset>

          <fieldset aria-describedby="fieldset2" data-original-aria-describedby="fieldset2">
            <input aria-describedby="input2" data-original-aria-describedby="input2">
            <textarea aria-describedby="textarea2" data-original-aria-describedby="textarea2"></textarea>
          </fieldset>
        </form>
      "
`;
